### PR TITLE
Refactor connection logic to allow more stateful message handling.

### LINKED ISSUE #232: Refactor message processing logic
I started trying to fill in more complex messages to our internal request/response protocol and identified a flaw in our state handling preventing us from combining multiple messages into a single response.

What I tried to do is test out more complex messages by adding a `Request::GetBlocksByHash` and a `Response::Blocks`: you assemble a collection of `BlockHeaderHash`es, send it as a `Request`, and get a collection of `Block`s back.  To do this, the peer connection has to turn the collection into a `Message::GetData` with corresponding `InventoryHash`es, then wait to collect individual `Message::Block`s until all of the requested blocks have been received.

However, it's not possible to do this right now, because the connection's state machine can only maintain the current request and a response channel. This means it has nowhere to accumulate the blocks themselves. In this example, the problem could be "fixed" by requesting one block at a time, but that doesn't solve the general problem, which is that we currently can't combine multiple messages into a single response.

To fix this, we should extract the message processing logic out of the current `handle_message_as_response` into a `Handler` that can accumulate state to combine multiple `Message`s into a single `Response` and eventually signal that the response is ready.

### PR DESCRIPTION
**Pull request by hdevalence**
Closes #232 .

This tests stateful message handling by requesting multiple blocks from the network.

### CODE REVIEW

File: zebra-chain/src/block.rs [hunk 1]
```diff
@@ -169,6 +169,12 @@ pub struct Block {
     pub transactions: Vec<Transaction>,
 }
 
+impl<'a> From<&'a Block> for BlockHeaderHash {
+    fn from(block: &'a Block) -> BlockHeaderHash {
+        (&block.header).into()
```
Comment (dconnolly) [inline zebra-chain/src/block.rs:174]: Why are the extra parens needed vs just `&block.header.into()`?

File: zebra-chain/src/transaction/transparent.rs [hunk 2]
```diff
@@ -24,17 +24,23 @@ pub struct OutPoint {
...
-    pub previous_output: OutPoint,
-
-    /// Computational Script for confirming transaction authorization.
-    pub signature_script: Script,
-
-    /// Transaction version as defined by the sender. Intended for
-    /// "replacement" of transactions when information is updated
-    /// before inclusion into a block.
-    pub sequence: u32,
+pub enum TransparentInput {
```
Comment (dconnolly) [inline zebra-chain/src/transaction/transparent.rs]: 👍

File: zebra-chain/src/transaction/serialize.rs [hunk 3]
```diff
@@ -34,20 +34,56 @@ impl ZcashDeserialize for OutPoint {
...
+            TransparentInput::PrevOut {
+                outpoint,
+                script,
+                sequence,
+            } => {
+                outpoint.zcash_serialize(&mut writer)?;
+                script.zcash_serialize(&mut writer)?;
+                writer.write_u32::<LittleEndian>(*sequence)?;
+            }
+            TransparentInput::Coinbase { data, sequence } => {
```
Comment (dconnolly) [inline zebra-chain/src/transaction/serialize.rs]: 💰 🔈

File: zebra-chain/src/transaction/serialize.rs [hunk 4]
```diff
@@ -34,20 +34,56 @@ impl ZcashDeserialize for OutPoint {
...
+                writer.write_all(&[0; 32][..])?;
+                writer.write_u32::<LittleEndian>(0xffff_ffff)?;
+                assert!(data.len() <= 100);
```
Comment (dconnolly) [inline zebra-chain/src/transaction/serialize.rs]: Why not then make data a [u8; 100]?

File: zebra-chain/src/transaction/serialize.rs [hunk 5]
```diff
@@ -34,20 +34,56 @@ impl ZcashDeserialize for OutPoint {
...
 }
 
 impl ZcashDeserialize for TransparentInput {
     fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
-        Ok(TransparentInput {
-            previous_output: OutPoint::zcash_deserialize(&mut reader)?,
-            signature_script: Script::zcash_deserialize(&mut reader)?,
-            sequence: reader.read_u32::<LittleEndian>()?,
-        })
+        // This inlines the OutPoint deserialization to peek at the hash value
```
Comment (dconnolly) [inline zebra-chain/src/transaction/serialize.rs]: 👍

**Review by dconnolly — APPROVED** [review]

File: zebra-network/src/protocol/internal/request.rs [hunk 6]
```diff
@@ -1,20 +1,24 @@
...
 
 /// A network request, represented in internal format.
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Clone, Debug)]
 pub enum Request {
     /// Requests additional peers from the server.
-    GetPeers,
+    Peers,
     /// Advertises peers to the remote server.
+    // XXX potentially remove this -- we don't use it?
```
Comment (dconnolly) [inline zebra-network/src/protocol/internal/request.rs]: 👍

File: zebra-network/src/peer/error.rs [hunk 7]
```diff
@@ -52,6 +52,9 @@ pub enum PeerError {
     // appropriate error when a `Reject` message is received.
     #[error("Received a Reject message")]
     Rejected,
+    /// The remote peer responded with a block we didn't ask for.
+    #[error("Remote peer responded with a block we didn't ask for.")]
+    WrongBlock,
```
Comment (dconnolly) [inline zebra-network/src/peer/error.rs:54]: 👍

File: zebra-network/src/peer/connection.rs [hunk 8]
```diff
@@ -391,14 +429,12 @@ where
...
-            Response::Transactions(txs) => {
-                let hashes = txs
-                    .into_iter()
-                    .map(|tx| InventoryHash::Tx(TransactionHash::from(tx)))
-                    .collect::<Vec<_>>();
-
-                if let Err(e) = self.peer_tx.send(Message::Inv(hashes)).await {
-                    self.fail_with(e.into());
+            Response::Blocks(blocks) => {
+                // Generate one block message per block.
```
Comment (dconnolly) [inline zebra-network/src/peer/connection.rs:417]: 👍

File: zebra-network/src/peer/connection.rs [hunk 9]
```diff
@@ -106,29 +174,53 @@ where
...
                         Either::Left((Some(Ok(peer_msg)), _timer)) => {
-                            match self.handle_message_as_response(peer_msg) {
-                                None => continue,
-                                Some(msg) => self.handle_message_as_request(msg).await,
+                            // Try to process the message using the handler.
+                            // This extremely awkward construction avoids
+                            // keeping a live reference to handler across the
+                            // call to handle_message_as_request, which takes
+                            // &mut self. This is a sign that we don't properly
+                            // factor the state required for inbound and
```
Comment (dconnolly) [inline zebra-network/src/peer/connection.rs:184]: 👍

File: zebra-chain/src/block.rs (hunk 1 above)
Comment (hdevalence) [reply zebra-chain/src/block.rs:174]: The problem is that `&block.header.into()` will make a `&T`, where `impl Into<T> for BlockHeader`, but we want `T` where `impl Into<T> for &BlockHeader`, so the & has to be inside the parentheses.

File: zebra-network/src/peer/connection.rs [hunk 10]
```diff
@@ -106,29 +174,53 @@ where
...
                             }
                         }
                         Either::Right(((), _peer_fut)) => {
                             trace!("client request timed out");
-                            // Re-matching lets us take ownership of tx
+                            let e = PeerError::ClientRequestTimeout;
                             self.state = match self.state {
-                                State::AwaitingResponse(Request::Ping(_), _) => {
-                                    self.fail_with(PeerError::ClientRequestTimeout);
+                                // Special case: ping timeouts fail the connection.
```
Comment (dconnolly) [inline zebra-network/src/peer/connection.rs:213]: 👍

File: zebra-network/src/peer/connection.rs [hunk 11]
```diff
@@ -106,29 +174,53 @@ where
...
+                                self.handle_message_as_request(msg).await;
+                            } else {
+                                // Otherwise, check whether the handler is finished
+                                // processing messages and update the state.
+                                self.state = match self.state {
+                                    State::AwaitingResponse(Handler::Finished(response), tx) => {
+                                        let _ = tx.send(response);
+                                        State::AwaitingRequest
+                                    }
+                                    pending @ State::AwaitingResponse(_, _) => pending,
```
Comment (dconnolly) [inline zebra-network/src/peer/connection.rs:204]: neat

File: zebra-network/src/peer/connection.rs [hunk 12]
```diff
@@ -9,24 +10,93 @@ use futures::{
...
+    fn process_message(&mut self, msg: Message) -> Option<Message> {
+        trace!(?msg);
+        // This function is where we statefully interpret Bitcoin/Zcash messages
+        // into responses to messages in the internal request/response protocol.
+        // This conversion is done by a sequence of (request, message) match arms,
+        // each of which contains the conversion logic for that pair.
+        use Handler::*;
+        let mut ignored_msg = None;
+        // XXX can this be avoided?
+        let mut tmp_state = std::mem::replace(self, Finished(Ok(Response::Ok)));
```
Comment (dconnolly) [inline zebra-network/src/peer/connection.rs]:
```suggestion
        let tmp_state = std::mem::replace(self, Finished(Ok(Response::Ok)));
```

**Review by dconnolly — APPROVED** [review]
🚚

File: zebra-network/src/protocol/internal/request.rs (hunk 6 above)
Comment (hdevalence) [reply zebra-network/src/protocol/internal/request.rs]: I can include this in this PR

**Pull request merged**